### PR TITLE
Fix clippy warnings

### LINKED ISSUE #3585: fix clippy -- -D warnings  ?
when run `cargo clippy -- -D warnings`, errors occur.

maybe we can fix it, and this rule to ci.yaml.

### PR DESCRIPTION
**Pull request by LinkTed**
## Motivation
To improve the code quality, the clippy warnings were fixed.

## Solution
Fix the clippy warnings. Also, prevent futures warnings to be merged by add the option `-D warnings` to the build job.

Fixes #3585.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -271,7 +271,7 @@ jobs:
 
       # Run clippy
       - name: "clippy --all"
-        run: cargo clippy --all --tests --all-features
+        run: cargo clippy --all --tests --all-features -- -D warnings
```
Comment (taiki-e) [inline .github/workflows/ci.yml:274]: This is not needed. We have already denied warnings (via RUSTFLAGS).

Comment (LinkTed) [reply .github/workflows/ci.yml:274]: Already fixed it :smile:.

Comment (taiki-e) [reply .github/workflows/ci.yml:274]: I don't think you've pushed that change.

**Comment by Darksonn** [comment]
The clippy setup doesn't find all the warnings because we fix clippy to a specific Rust version. This is because otherwise PRs randomly fail due unrelated clippy errors whenever new lints are added. We are always happy to fix clippy errors from newer versions of clippy.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Darksonn) [reply .github/workflows/ci.yml:274]:
```suggestion
        run: cargo clippy --all --tests --all-features
```

Comment (LinkTed) [reply .github/workflows/ci.yml]: Sorry. I forgot to push it.

**Comment by LinkTed** [comment]
@Darksonn Thanks, but why did the build job fail?

**Comment by Darksonn** [comment]
The CI failure will be fixed in  #4020. You can either wait for that to be merged, or rebase this PR on an earlier commit in the master branch (e.g. `cf02b3f32d12223283d475f65ccf9301a9a2b951`).

**Comment by Darksonn** [comment]
The PR has been merged.

**Comment by LinkTed** [comment]
@Darksonn Thanks for the notification :smile:.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**